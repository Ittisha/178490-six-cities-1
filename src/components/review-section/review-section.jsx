import * as React from 'react';
import {connect} from 'react-redux';

import {ReviewsList} from '../reviews-list/reviews-list';
import {getIsAuthorized, getReviews} from '../../reducers/selectors';
import {Operation} from '../../reducers/review/review';
import PropTypes from 'prop-types';
import userPropTypes from '../../props/user';
import {MAX_REVIEWS_NUMBER} from '../../consts';
import {withFormSubmit} from '../../hoc/with-form-submit/with-form-submit';
import ReviewForm from '../review-form/review-form';

const ReviewFormWithFormSubmit = withFormSubmit(ReviewForm);

class ReviewSection extends React.PureComponent {
  componentDidMount() {
    const {loadReviews, offerId} = this.props;
    loadReviews(offerId);
  }

  render() {
    const {
      reviews,
      offerId,
      isAuthorized,
    } = this.props;

    const hasReviewsToShow = reviews && reviews.length !== 0;

    const reviewsToRender = reviews.slice(0, MAX_REVIEWS_NUMBER);

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
        {hasReviewsToShow && <ReviewsList reviews={reviewsToRender} />}
        {isAuthorized && <ReviewFormWithFormSubmit offerId={offerId}/>}
      </section>
    );
  }
}

ReviewSection.propTypes = {
  reviews: PropTypes.arrayOf(PropTypes.shape({
    user: userPropTypes,
    rating: PropTypes.number,
    comment: PropTypes.string,
    date: PropTypes.string,
  })),
  offerId: PropTypes.number.isRequired,
  loadReviews: PropTypes.func.isRequired,
  isAuthorized: PropTypes.bool.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  reviews: getReviews(state),
  isAuthorized: getIsAuthorized(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadReviews: (id) => dispatch(Operation.loadReviews(id))
});

export {ReviewSection};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewSection);
