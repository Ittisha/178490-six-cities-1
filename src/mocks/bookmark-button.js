export const MOCK_PROPS = {
  offerId: 1,
  isInBookmarks: false,
  className: ``,
  width: 15,
  height: 12,
};
