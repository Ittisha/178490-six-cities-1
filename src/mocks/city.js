export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
  zoom: 12,
};

export const MOCK_CITIES = [`Amsterdam`, `Paris`];

export const MOCK_CITIES_COORDS = {
  Amsterdam: [52.38333, 4.9],
  Paris: [48.864716, 2.349014],
};

export const MOCK_CITIES_ZOOM = {
  Amsterdam: 17,
  Paris: 15,
};
