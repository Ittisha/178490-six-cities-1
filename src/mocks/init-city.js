export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
};
