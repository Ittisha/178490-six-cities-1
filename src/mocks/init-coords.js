export const INIT_COORDS = [52.38333, 4.9];
