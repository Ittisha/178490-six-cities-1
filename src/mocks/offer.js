export const OFFER = {
  "city": {
    "name": `Cologne`,
    "location": {
      "latitude": 50.938361,
      "longitude": 6.959974,
      "zoom": 13}},
  "preview_image": `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
  "images": [
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": false,
  "rating": 2.5,
  "type": `house`,
  "bedrooms": 2,
  "max_adults": 6,
  "price": 757,
  "goods": [
    `Coffee machine`,
    `Baby seat`,
    `Dishwasher`,
    `Fridge`,
    `Washing machine`,
    `Breakfast`,
  ],
  "host": {
    "id": 25,
    "name": `Angelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-angelina.jpg`
  },
  "description": `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  "location": {
    "latitude": 50.945361,
    "longitude": 6.935974,
    "zoom": 16
  },
  "id": 1
};
