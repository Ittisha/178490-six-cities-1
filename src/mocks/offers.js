export const OFFERS = [
  {
    isPremium: true,
    isInBookmarks: false,
    title: `Beautiful & luxurious apartment at great location`,
    photoUrl: `img/apartment-01.jpg`,
    price: 120,
    type: `Apartment`,
    rating: `93%`,
    id: 1,
    coordinates: [52.3909553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: true,
    title: `Wood and stone place`,
    photoUrl: `img/room.jpg`,
    price: 80,
    type: `Private room`,
    rating: `80%`,
    id: 2,
    coordinates: [52.369553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Canal View Prinsengracht`,
    photoUrl: `img/apartment-02.jpg`,
    price: 132,
    type: `Apartment`,
    rating: `80%`,
    id: 3,
    coordinates: [52.3909553943508, 4.929309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Nice, cozy, warm big bed apartment`,
    photoUrl: `img/apartment-03.jpg`,
    price: 180,
    type: `Apartment`,
    rating: `80%`,
    id: 4,
    coordinates: [52.3809553943508, 4.939309666406198],
  }
];
