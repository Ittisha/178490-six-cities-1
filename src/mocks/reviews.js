export const MOCK_REVIEWS = [
  {
    id: 1,
    user: {
      avatarUrl: ``,
      id: 1,
      isPro: false,
      name: ``,
    },
    rating: 4,
    comment: ``,
    date: ``,
  },
  {
    id: 2,
    user: {
      avatarUrl: ``,
      id: 5,
      isPro: false,
      name: ``,
    },
    rating: 3,
    comment: ``,
    date: ``,
  }
];

export const MOCK_REVIEW = {
  comment: ``,
  id: 1,
  rating: `3`,
};
