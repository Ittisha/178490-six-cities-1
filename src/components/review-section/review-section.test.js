import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {ReviewSection} from './review-section';
import {MOCK_REVIEWS} from '../../mocks/reviews';

it(`Review correctly renders`, () => {
  const renderer = new ShallowRenderer();
  const tree = renderer
    .render(<ReviewSection
      offerId={1}
      loadReviews={() => {}}
      reviews={MOCK_REVIEWS}
      isAuthorized={true}
    />);

  expect(tree).toMatchSnapshot();
});
