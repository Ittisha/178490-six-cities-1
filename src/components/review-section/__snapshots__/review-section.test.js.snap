// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review correctly renders 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ReviewsList
    reviews={
      Array [
        Object {
          "comment": "",
          "date": "",
          "id": 1,
          "rating": 4,
          "user": Object {
            "avatarUrl": "",
            "id": 1,
            "isPro": false,
            "name": "",
          },
        },
        Object {
          "comment": "",
          "date": "",
          "id": 2,
          "rating": 3,
          "user": Object {
            "avatarUrl": "",
            "id": 5,
            "isPro": false,
            "name": "",
          },
        },
      ]
    }
  />
  <WithFormSubmit
    offerId={1}
  />
</section>
`;
